runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/endpoints/util/Util.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package aws.smithy.kotlin.runtime.http.endpoints.util
+
+import aws.smithy.kotlin.runtime.util.InternalApi
+import aws.smithy.kotlin.runtime.util.net.Host
+import aws.smithy.kotlin.runtime.util.net.isIpv6
+import aws.smithy.kotlin.runtime.util.net.isValidHostname
+import aws.smithy.kotlin.runtime.util.text.ensureSuffix
+import aws.smithy.kotlin.runtime.util.text.urlEncodeComponent
+
+@InternalApi
+public fun isSet(value: Any?): Boolean = value != null
+
+@InternalApi
+public fun not(value: Boolean): Boolean = !value
+
+@InternalApi
+public fun substring(value: String, start: Int, stop: Int, reverse: Boolean): String? =
+    when {
+        start >= stop || stop > value.length -> null
+        reverse -> value.substring(value.length - stop until value.length - start)
+        else -> value.substring(start until stop)
+    }
+
+@InternalApi
+public fun stringEquals(i: String, j: String): Boolean = i == j
+
+@InternalApi
+public fun isValidHostLabel(value: String, allowSubdomains: Boolean): Boolean =
+    if (allowSubdomains) value.isValidHostname() else value.split('.').all(String::isValidHostname)
+
+@InternalApi
+public fun uriEncode(value: String): String = value.urlEncodeComponent(formUrlEncode = false)
+
+@InternalApi
+public fun booleanEquals(i: Boolean, j: Boolean): Boolean = i == j
```

Comment:
where is this used? Could probably be `inline` (in fact several of these could be)

Reply:
I completely agree, but apparently the compiler doesn't like it:
`Expected performance impact from inlining is insignificant. Inlining works best for functions with parameters of functional types`.

Reply:
My next question was going to be do some of these _need_ to be actual functions or can they be dealt with in codegen directly?